File: services/geo/geo/core_geometry.py
"""The bounded planar/prism kernel; no native parser, inferred geometry or repairs."""
import math
from .core_contract import bundled_schema, fail, index_records, parse_core, ref_key, require_revision, version_key
from .core_identity import validate_core_identity_graph
from .core_sources import is_precise_core_locator, unique, validate_core_source_catalog
from .core_frames import _horizontal, validate_core_frame_catalog


def _inline(rep):
    profile = rep["geometry"]
    return profile["geometry"] if profile["profile"] == "planar" else profile["footprint"] if profile["profile"] == "prism" else None


def _polygons(g):
    return [g["coordinates"]] if g["type"] == "Polygon" else g["coordinates"] if g["type"] == "MultiPolygon" else []


def _points(g):
    if g["type"] == "Point":
        return [g["coordinates"]]
    if g["type"] == "LineString":
        return g["coordinates"]
    return [point for polygon in _polygons(g) for ring in polygon for point in ring]


def _normalise(g, frame, factors):
    def point(p):
        return list(_horizontal(frame, p, factors))
    if g["type"] == "Point":
        coords = point(g["coordinates"])
    elif g["type"] == "LineString":
        coords = [point(p) for p in g["coordinates"]]
    elif g["type"] == "Polygon":
        coords = [[point(p) for p in ring] for ring in g["coordinates"]]
    else:
        coords = [[[point(p) for p in ring] for ring in polygon] for polygon in g["coordinates"]]
    return {"type": g["type"], "coordinates": coords}


def _ring_area(ring):
    ox, oy = ring[0]
    total = 0
    for a, b in zip(ring, ring[1:]):
        total += (a[0] - ox) * (b[1] - oy) - (b[0] - ox) * (a[1] - oy)
    return total / 2


def _polygon_area(p):
    return abs(_ring_area(p[0])) - sum(abs(_ring_area(r)) for r in p[1:])


def _topology_work(g):
    polygons = _polygons(g)
    total = 0
    for polygon in polygons:
        for i, ring in enumerate(polygon):
            n = len(ring) - 1
            total += n * (n - 3) // 2
            total += sum(n * (len(other) - 1) for other in polygon[:i])
    for i, polygon in enumerate(polygons):
        total += sum(sum(len(r) - 1 for r in polygon) * sum(len(r) - 1 for r in other) for other in polygons[:i])
    return total


def _topology_issue(g, epsilon):
    def orient(a, b, c):
        return (b[0] - a[0]) * (c[1] - a[1]) - (b[1] - a[1]) * (c[0] - a[0])

    def on_segment(a, b, p):
        return (abs(orient(a, b, p)) <= epsilon
                and min(a[0], b[0]) - epsilon <= p[0] <= max(a[0], b[0]) + epsilon
                and min(a[1], b[1]) - epsilon <= p[1] <= max(a[1], b[1]) + epsilon)

    def touches(a, b, c, d):
        u, v, w, x = orient(a, b, c), orient(a, b, d), orient(c, d, a), orient(c, d, b)
        crossing = ((u > epsilon and v < -epsilon or u < -epsilon and v > epsilon)
                    and (w > epsilon and x < -epsilon or w < -epsilon and x > epsilon))
        return crossing or on_segment(a, b, c) or on_segment(a, b, d) or on_segment(c, d, a) or on_segment(c, d, b)

    def contains(ring, p):
        inside = False
        for a, b in zip(ring, ring[1:]):
            if on_segment(a, b, p):
                return 0
            if (a[1] > p[1]) != (b[1] > p[1]) and p[0] < (b[0] - a[0]) * (p[1] - a[1]) / (b[1] - a[1]) + a[0]:
                inside = not inside
        return 1 if inside else -1

    def rings_touch(a, b):
        return any(touches(a[i], a[i + 1], b[j], b[j + 1]) for i in range(len(a) - 1) for j in range(len(b) - 1))

    def in_polygon(rings, p):
        return contains(rings[0], p) == 1 and not any(contains(r, p) >= 0 for r in rings[1:])

    polygons = _polygons(g)
    for polygon in polygons:
        for ring in polygon:
            n = len(ring) - 1
            for i in range(n):
                if math.hypot(ring[i + 1][0] - ring[i][0], ring[i + 1][1] - ring[i][1]) <= epsilon:
                    return "Repeated polygon vertex"
                for j in range(i + 1, n):
                    if j == i + 1 or i == 0 and j == n - 1:
                        continue
                    if touches(ring[i], ring[i + 1], ring[j], ring[j + 1]):
                        return "Self-intersecting or self-touching ring"
        for h in range(1, len(polygon)):
            if contains(polygon[0], polygon[h][0]) != 1 or rings_touch(polygon[0], polygon[h]):
                return "Hole is not strictly inside shell"
            for k in range(1, h):
                if rings_touch(polygon[k], polygon[h]) or contains(polygon[k], polygon[h][0]) >= 0 or contains(polygon[h], polygon[k][0]) >= 0:
                    return "Overlapping or nested holes"
    for a in range(len(polygons)):
        for b in range(a + 1, len(polygons)):
            if any(rings_touch(r, s) for r in polygons[a] for s in polygons[b]):
                return "Touching multipart boundaries require worker validation"
            if in_polygon(polygons[a], polygons[b][0][0]) or in_polygon(polygons[b], polygons[a][0][0]):
                return "Overlapping polygon parts"
    return None


def _prepare(value, identity_value, sources_value, frames_value):
    identity = validate_core_identity_graph(identity_value)
    sources = validate_core_source_catalog(sources_value, identity)
    frames = validate_core_frame_catalog(frames_value)
    catalog = parse_core("geometry-catalog", value)
    policy = bundled_schema("geometry-catalog")[0]["x-ulpin-geometry-policy"]
    factors = bundled_schema("frame-catalog")[0]["x-ulpin-frame-policy"]["lengthMetres"]
    entities = index_records(identity["entities"], "entity")
    frame_index = index_records(frames["frames"], "frame")
    parts, assets = index_records(sources["parts"], "source part"), index_records(sources["assets"], "asset")
    reps = index_records(catalog["representations"], "representation")
    reported = index_records(catalog["reportedQuantities"], "reported quantity")
    positions, comparisons = 0, 0
    for rep in catalog["representations"]:
        if ref_key(rep["entity"]) not in entities:
            fail("MISSING_TARGET", "Representation entity is absent")
        frame = require_revision(frame_index, rep["frame"], "representation frame") if rep["frame"] else None
        unique([version_key(link) for link in rep["sourceParts"]], "source parts")
        for link in rep["sourceParts"]:
            require_revision(parts, link, "representation source part")
        g = _inline(rep)
        if g:
            for link in rep["sourceParts"]:
                part = require_revision(parts, link, "geometry source part")
                if not any(is_precise_core_locator(locator) for locator in part["locators"]):
                    fail("LOCATOR_NOT_QUALIFIED", "Inline geometry needs an exact source locator")
            positions += len(_points(g))
            if positions > policy["maximumPositions"]:
                fail("GEOMETRY_BUDGET", "Catalog exceeds inline position profile")
            for polygon in _polygons(g):
                for ring in polygon:
                    if ring[0] != ring[-1]:
                        fail("GEOMETRY_RING", "Polygon rings must be explicitly closed")
            if frame and frame["kind"] == "engineering":
                metric = _normalise(g, frame, factors)
                comparisons += _topology_work(metric)
                if comparisons > policy["maximumTopologyComparisons"]:
                    fail("GEOMETRY_BUDGET", "Catalog exceeds bounded topology comparisons")
                if not all(math.isfinite(v) for p in _points(metric) for v in p):
                    fail("GEOMETRY_NUMERIC", "Metre normalization overflowed")
                for polygon in _polygons(metric):
                    for ring in polygon:
                        area = abs(_ring_area(ring))
                        if not math.isfinite(area) or area <= policy["areaFloorMetres2"]:
                            fail("GEOMETRY_DEGENERATE", "Ring has no qualified positive area")
                if _polygons(metric):
                    issue = _topology_issue(metric, policy["epsilonMetres"])
                    if issue:
                        fail("GEOMETRY_TOPOLOGY", issue)
                elif metric["type"] == "LineString" and all(p == metric["coordinates"][0] for p in metric["coordinates"]):
                    fail("GEOMETRY_DEGENERATE", "Line has no distinct positions")
        profile = rep["geometry"]
        if profile["profile"] == "asset":
            require_revision(assets, profile["asset"], "geometry asset")
        if profile["profile"] == "prism" and profile["interval"] is not None:
            interval = profile["interval"]
            if interval["upperMetres"] < interval["lowerMetres"]:
                fail("GEOMETRY_INTERVAL", "Positive-up metre elevations are reversed")
            if not math.isfinite(interval["upperMetres"] - interval["lowerMetres"]):
                fail("GEOMETRY_NUMERIC", "Vertical extent overflowed")
            if not frame or frame["kind"] == "geocentric" or frame["vertical"]["kind"] not in ("benchmark", "datum"):
                fail("VERTICAL_UNRESOLVED", "Prism requires its exact named vertical reference")
            if version_key(interval["reference"]) != version_key(frame["vertical"]["reference"]):
                fail("VERTICAL_MISMATCH", "Prism bounds do not use frame vertical reference")
    for quantity in catalog["reportedQuantities"]:
        if ref_key(quantity["entity"]) not in entities:
            fail("MISSING_TARGET", "Reported quantity entity is absent")
        require_revision(parts, quantity["sourcePart"], "reported quantity source part")
        if quantity["unit"] not in policy["reportedUnits"][quantity["definition"]]:
            fail("QUANTITY_UNIT", "Reported definition and unit dimensions differ")
        if quantity["amount"]["state"] == "conflicting":
            unique([version_key(link) for link in quantity["amount"]["candidates"]], "quantity candidates")
            values = set()
            for link in quantity["amount"]["candidates"]:
                other = require_revision(reported, link, "quantity candidate")
                if (ref_key(other["entity"]) != ref_key(quantity["entity"]) or other["definition"] != quantity["definition"]
                        or other["unit"] != quantity["unit"] or other["amount"]["state"] != "known"):
                    fail("QUANTITY_CONFLICT", "Conflict candidates must be known comparable quantities")
                values.add(other["amount"]["value"])
            if len(values) < 2:
                fail("QUANTITY_CONFLICT", "Equal reported values do not form a numeric disagreement")
    return catalog, entities, frame_index, reps, policy, factors


def validate_core_geometry_catalog(value, identity, sources, frames):
    return _prepare(value, identity, sources, frames)[0]


def evaluate_core_geometry(value, identity, sources, frames):
    catalog, entities, frame_index, _, policy, factors = _prepare(value, identity, sources, frames)
    def available(value, reason):
        return {"available": bool(value), "reasonCode": None if value else reason}
    result = []
    for rep in catalog["representations"]:
        frame = require_revision(frame_index, rep["frame"], "frame") if rep["frame"] else None
        g = _inline(rep)
        polygon = bool(g and _polygons(g))
        profile = rep["geometry"]
        interval = profile["interval"] if profile["profile"] == "prism" else None
        horizontal = _compute_quantity(rep, frame, "planar_length" if g and g["type"] == "LineString" else "horizontal_area", policy, factors)
        volume = _compute_quantity(rep, frame, "prism_volume", policy, factors)
        kind = entities[ref_key(rep["entity"])]["kind"]
        caps = {
            "source_reference": available(bool(rep["sourceParts"]), "NO_SOURCE_PART"),
            "local_preview": available(g is not None, "NO_INLINE_GEOMETRY"),
            "horizontal_measurement": available(horizontal["value"] is not None, horizontal["reasonCode"] or "QUANTITY_GEOMETRY"),
            "prism_volume": available(volume["value"] is not None, volume["reasonCode"] or "QUANTITY_GEOMETRY_ROLE"),
            "exterior_render": available(volume["value"] is not None and interval is not None and interval["upperMetres"] > interval["lowerMetres"], "NO_POSITIVE_QUALIFIED_PRISM"),
            "interior_selection": available(polygon and (kind == "space" and rep["role"] == "unit_boundary" or kind == "level" and rep["role"] == "floor_boundary"), "NO_IDENTIFIED_INTERIOR"),
        }
        result.append({"representation": {"ref": rep["ref"], "revision": rep["revision"]}, "capabilities": caps})
    return result


def measure_core_representation(value, identity, sources, frames, request_value):
    _, _, frame_index, reps, policy, factors = _prepare(value, identity, sources, frames)
    request = parse_core("measure-request", request_value)
    rep = require_revision(reps, request["representation"], "measurement representation")
    frame = require_revision(frame_index, rep["frame"], "frame") if rep["frame"] else None
    definition = request["definition"]
    base = {"definition": definition, "unit": {"horizontal_area": "m2", "prism_volume": "m3", "planar_length": "m"}[definition],
            "representation": request["representation"], "frame": rep["frame"], "method": policy["method"], "sourceParts": rep["sourceParts"]}
    return {**base, **_compute_quantity(rep, frame, definition, policy, factors)}


def _compute_quantity(rep, frame, definition, policy, factors):
    g = _inline(rep)
    def unavailable(reason):
        return {"value": None, "reasonCode": reason}
    if not g:
        return unavailable("NO_INLINE_GEOMETRY")
    if not frame or frame["kind"] != "engineering":
        return unavailable("ENGINEERING_FRAME_REQUIRED")
    if rep["role"] in policy["nonAnalyticalRoles"]:
        return unavailable("NON_ANALYTICAL_ROLE")
    metric = _normalise(g, frame, factors)
    if definition == "planar_length":
        if metric["type"] != "LineString":
            return unavailable("LINE_REQUIRED")
        result = sum(math.hypot(b[0] - a[0], b[1] - a[1]) for a, b in zip(metric["coordinates"], metric["coordinates"][1:]))
    else:
        polygons = _polygons(metric)
        if not polygons:
            return unavailable("POLYGON_REQUIRED")
        result = sum(_polygon_area(p) for p in polygons)
        if definition == "prism_volume":
            if rep["role"] not in policy["volumeRoles"]:
                return unavailable("QUANTITY_GEOMETRY_ROLE")
            if rep["geometry"]["profile"] != "prism" or rep["geometry"]["interval"] is None:
                return unavailable("UNKNOWN_VERTICAL_INTERVAL")
            interval = rep["geometry"]["interval"]
            result *= interval["upperMetres"] - interval["lowerMetres"]
    if not math.isfinite(result) or result < 0:
        return unavailable("QUANTITY_NUMERIC")
    return {"value": result, "reasonCode": None}

File: tests/fixtures/core-geometry.ts
import {entity,ref} from "./core-identity";
import {sourceFixture} from "./core-sources";
import {frameFixture,frameRef} from "./core-frames";

export const representationRef=(id="exterior",revision=1)=>({ref:{namespace:"representation" as const,id},revision});
export function geometryFixture() {
  const source=sourceFixture();
  return {
    identity:{entities:[entity("B1"),entity("U1","space"),entity("L1","level")],relations:[]},
    sources:{...source,links:[],assets:source.assets.map(a=>({...a,mediaType:"application/geo+json"})),parts:source.parts.map(p=>({...p,locators:[{kind:"feature" as const,featureId:"B1"}]}))},
    frames:{frames:[frameFixture().frames[1]],operations:[]},
    geometry:{representations:[{
      ref:{namespace:"representation" as const,id:"exterior"},revision:1,entity:ref("B1"),frame:frameRef("metric"),role:"exterior" as const,
      sourceParts:[{ref:{namespace:"source_part" as const,id:"p1"},revision:1}],
      geometry:{profile:"prism" as const,footprint:{type:"Polygon" as const,coordinates:[[[0,0],[10,0],[10,8],[0,8],[0,0]]]},interval:{lowerMetres:0,upperMetres:6,reference:{ref:{namespace:"benchmark",id:"target-zero"},revision:1}}},
    }],reportedQuantities:[]},
    request:{representation:representationRef(),definition:"horizontal_area" as const},
  };
}
export interface GeometryCase {id:string;input:unknown;valid:boolean;value?:number|null;reasonCode?:string|null;code?:string;capabilities?:Record<string,boolean>}
export function geometryCases():GeometryCase[] {
  const cases:GeometryCase[]=[];
  const good=(id:string,value:number|null,change:(x:any)=>void=()=>{},reasonCode:string|null=null,capabilities?:Record<string,boolean>)=>{
    const input=JSON.parse(JSON.stringify(geometryFixture()));change(input);cases.push({id,input,valid:true,value,reasonCode,capabilities});
  };
  const bad=(id:string,code:string,change:(x:any)=>void)=>{const input=JSON.parse(JSON.stringify(geometryFixture()));change(input);cases.push({id,input,valid:false,code});};
  const rep=(x:any)=>x.geometry.representations[0];
  good("independent-rectangle-area",80);
  good("existing-revision-zero-is-preserved",80,x=>{rep(x).revision=0;x.request.representation.revision=0;});
  good("independent-six-metre-volume",480,x=>x.request.definition="prism_volume",null,{prism_volume:true,interior_selection:false});
  good("courtyard-subtraction",76,x=>rep(x).geometry.footprint.coordinates.push([[2,2],[4,2],[4,4],[2,4],[2,2]]));
  good("concave-outline",32,x=>rep(x).geometry.footprint.coordinates=[[[0,0],[10,0],[10,2],[2,2],[2,8],[0,8],[0,0]]]);
  good("disjoint-multipart-sum",86,x=>{rep(x).geometry.footprint={type:"MultiPolygon",coordinates:[rep(x).geometry.footprint.coordinates,[[[20,0],[22,0],[22,3],[20,3],[20,0]]]]};});
  good("ring-order-does-not-change-area",80,x=>rep(x).geometry.footprint.coordinates[0].reverse());
  good("far-from-origin-keeps-area",80,x=>rep(x).geometry.footprint.coordinates[0]=rep(x).geometry.footprint.coordinates[0].map((p:number[])=>p.map(n=>n+1e9)));
  good("millimetres-are-normalized-before-area",80,x=>{x.frames.frames[0].horizontalUnit="mm";rep(x).geometry.footprint.coordinates[0]=rep(x).geometry.footprint.coordinates[0].map((p:number[])=>p.map(n=>n*1000));});
  good("feet-are-not-square-metres",7.4322432,x=>x.frames.frames[0].horizontalUnit="ft");
  good("known-zero-volume-is-not-missing",0,x=>{x.request.definition="prism_volume";rep(x).geometry.interval.upperMetres=0;},null,{prism_volume:true,exterior_render:false});
  good("three-four-five-and-three-length",8,x=>{x.request.definition="planar_length";rep(x).role="alignment";rep(x).geometry={profile:"planar",geometry:{type:"LineString",coordinates:[[0,0],[3,4],[6,4]]}};});
  good("unknown-height-does-not-block-area",80,x=>rep(x).geometry.interval=null,null,{horizontal_measurement:true,prism_volume:false});
  good("unknown-height-blocks-volume",null,x=>{rep(x).geometry.interval=null;x.request.definition="prism_volume";},"UNKNOWN_VERTICAL_INTERVAL");
  good("unknown-datum-does-not-block-area",80,x=>{rep(x).geometry.interval=null;x.frames.frames[0].vertical={kind:"unknown",reason:"No datum"};});
  good("unplaced-geometry-remains-previewable",null,x=>{rep(x).frame=null;rep(x).geometry.interval=null;},"ENGINEERING_FRAME_REQUIRED",{local_preview:true,horizontal_measurement:false});
  good("display-is-not-analysis",null,x=>rep(x).role="display_only","NON_ANALYTICAL_ROLE");
  good("unspecified-is-not-analysis",null,x=>rep(x).role="unspecified","NON_ANALYTICAL_ROLE");
  good("roof-outline-does-not-prove-volume",null,x=>{rep(x).role="roof_projection";x.request.definition="prism_volume";},"QUANTITY_GEOMETRY_ROLE");
  good("native-asset-does-not-prove-solid",null,x=>rep(x).geometry={profile:"asset",asset:{ref:{namespace:"asset",id:"a1"},revision:1},format:"native"},"NO_INLINE_GEOMETRY",{prism_volume:false,local_preview:false});
  good("unavailable-is-not-zero",null,x=>{rep(x).geometry={profile:"unavailable",reason:"Registry without geometry"};rep(x).frame=null;},"NO_INLINE_GEOMETRY");
  good("identified-unit-is-selectable",80,x=>{rep(x).entity=ref("U1");rep(x).role="unit_boundary";},null,{interior_selection:true});
  good("exterior-does-not-invent-interiors",80,()=>{},null,{interior_selection:false});
  good("zero-document-geometry",80,x=>{x.sources={datasets:[],assets:[],sources:[],parts:[],links:[]};rep(x).sourceParts=[];});
  good("point-is-not-area",null,x=>rep(x).geometry={profile:"planar",geometry:{type:"Point",coordinates:[0,0]}},"POLYGON_REQUIRED");
  good("overflow-is-not-a-plausible-quantity",null,x=>{rep(x).geometry.footprint.coordinates[0]=rep(x).geometry.footprint.coordinates[0].map((p:number[])=>p.map(n=>n*1e110));rep(x).geometry.interval.upperMetres=1e110;x.request.definition="prism_volume";},"QUANTITY_NUMERIC",{prism_volume:false,exterior_render:false,horizontal_measurement:true});
  bad("unknown-entity","MISSING_TARGET",x=>rep(x).entity=ref("missing"));
  bad("stale-frame","STALE_REFERENCE",x=>rep(x).frame.revision=2);
  bad("stale-part","STALE_REFERENCE",x=>rep(x).sourceParts[0].revision=2);
  bad("vague-source-is-not-geometry-evidence","LOCATOR_NOT_QUALIFIED",x=>x.sources.parts[0].locators=[{kind:"verbatim",locator:"Somewhere in the file"}]);
  bad("duplicate-part","DUPLICATE_REFERENCE",x=>rep(x).sourceParts.push(rep(x).sourceParts[0]));
  bad("duplicate-representation","DUPLICATE_RECORD",x=>x.geometry.representations.push(rep(x)));
  bad("stale-measurement-revision","STALE_REFERENCE",x=>x.request.representation.revision=2);
  bad("planar-Z-cannot-be-dropped","INVALID_CONTRACT",x=>rep(x).geometry.footprint.coordinates[0][0].push(1));
  bad("unclosed-polygon","GEOMETRY_RING",x=>rep(x).geometry.footprint.coordinates[0][4]=[1,1]);
  bad("collinear-polygon","GEOMETRY_DEGENERATE",x=>rep(x).geometry.footprint.coordinates=[[[0,0],[1,0],[2,0],[0,0]]]);
  bad("asymmetric-bow-tie","GEOMETRY_TOPOLOGY",x=>rep(x).geometry.footprint.coordinates=[[[0,0],[10,8],[0,8],[8,0],[0,0]]]);
  bad("hole-outside-shell","GEOMETRY_TOPOLOGY",x=>rep(x).geometry.footprint.coordinates.push([[20,20],[22,20],[22,22],[20,22],[20,20]]));
  bad("overlapping-holes","GEOMETRY_TOPOLOGY",x=>rep(x).geometry.footprint.coordinates.push([[2,2],[6,2],[6,6],[2,6],[2,2]],[[4,4],[8,4],[8,7],[4,7],[4,4]]));
  bad("overlapping-parts","GEOMETRY_TOPOLOGY",x=>{const p=rep(x).geometry.footprint.coordinates;rep(x).geometry.footprint={type:"MultiPolygon",coordinates:[p,p]};});
  bad("reversed-positive-up-bounds","GEOMETRY_INTERVAL",x=>rep(x).geometry.interval.upperMetres=-1);
  bad("prism-without-datum","VERTICAL_UNRESOLVED",x=>x.frames.frames[0].vertical={kind:"unknown",reason:"Absent"});
  bad("wrong-prism-benchmark","VERTICAL_MISMATCH",x=>rep(x).geometry.interval.reference.ref.id="unrelated");
  bad("topology-work-is-bounded","GEOMETRY_BUDGET",x=>{const ring=Array.from({length:1000},(_,i)=>[Math.cos(i*2*Math.PI/1000),Math.sin(i*2*Math.PI/1000)]);ring.push(ring[0]);rep(x).geometry.footprint.coordinates=[ring];});
  const reported=(x:any)=>{x.geometry.reportedQuantities=[{ref:{namespace:"reported_quantity",id:"q1"},revision:1,entity:ref("B1"),definition:"gross_floor_area",unit:"m2",amount:{state:"known",value:100},sourcePart:{ref:{namespace:"source_part",id:"p1"},revision:1}}];};
  good("reported-gross-area-does-not-replace-footprint",80,reported);
  bad("reported-unit-mismatch","QUANTITY_UNIT",x=>{reported(x);x.geometry.reportedQuantities[0].unit="m3";});
  bad("reported-value-negative","INVALID_CONTRACT",x=>{reported(x);x.geometry.reportedQuantities[0].amount.value=-1;});
  const conflict=(x:any)=>{reported(x);const first=x.geometry.reportedQuantities[0],second={...first,ref:{namespace:"reported_quantity",id:"q2"},amount:{state:"known",value:120}};x.geometry.reportedQuantities.push(second,{...first,ref:{namespace:"reported_quantity",id:"q3"},amount:{state:"conflicting",candidates:[{ref:first.ref,revision:1},{ref:second.ref,revision:1}],reason:"Two sources disagree"}});};
  good("reported-disagreement-is-retained-separately",80,conflict);
  bad("conflict-cannot-mix-net-and-gross","QUANTITY_CONFLICT",x=>{conflict(x);x.geometry.reportedQuantities[1].definition="net_floor_area";});
  bad("conflict-cannot-use-stale-value","STALE_REFERENCE",x=>{conflict(x);x.geometry.reportedQuantities[2].amount.candidates[1].revision=2;});
  bad("conflict-cannot-repeat-one-value","DUPLICATE_REFERENCE",x=>{conflict(x);x.geometry.reportedQuantities[2].amount.candidates[1]=x.geometry.reportedQuantities[2].amount.candidates[0];});
  bad("conflict-cannot-reference-itself","QUANTITY_CONFLICT",x=>{conflict(x);x.geometry.reportedQuantities[2].amount.candidates[0]={ref:{namespace:"reported_quantity",id:"q3"},revision:1};});
  bad("equal-quantities-are-not-a-disagreement","QUANTITY_CONFLICT",x=>{conflict(x);x.geometry.reportedQuantities[1].amount.value=100;});
  return cases;
}
